### PR TITLE
[NVIDIA XLA] Added a helper function to assign suffix to instruction name when cloning

### PR DESCRIPTION
**Pull request by Tixxx**
CloneWithNewOperands function doesnt add suffix to the cloned instruction even suffix is specified in the context.
Reuse existing logic to add suffix.

### CODE REVIEW

File: xla/hlo/ir/hlo_instruction.cc [hunk 1]
```diff
@@ -45,6 +45,13 @@ limitations under the License.
 #include "absl/strings/str_join.h"
 #include "absl/strings/string_view.h"
 #include "absl/types/span.h"
+#include "tsl/lib/gtl/iterator_range.h"
+#include "tsl/lib/gtl/map_util.h"
+#include "tsl/platform/errors.h"
+#include "tsl/platform/human_readable_json.h"
+#include "tsl/platform/logging.h"  // IWYU pragma: keep
+#include "tsl/platform/status.h"
+#include "tsl/platform/statusor.h"
```
Comment (cheshire) [inline xla/hlo/ir/hlo_instruction.cc:54]: Do we need to move all those includes?

File: xla/hlo/ir/hlo_instruction.cc [hunk 2]
```diff
@@ -2074,6 +2074,38 @@ bool HloInstruction::HasSideEffect() const {
          execution_threads_set.contains(execution_thread);
 }
 
+void HloInstruction::AddSuffixToInstructionName(const std::string& suffix) {
```
Comment (cheshire) [inline xla/hlo/ir/hlo_instruction.cc:2077]: Nit: absl::string_view?

File: xla/hlo/ir/hlo_instruction.cc [hunk 3]
```diff
@@ -2074,6 +2074,38 @@ bool HloInstruction::HasSideEffect() const {
...
+  // If an instruction is cloned multiple times avoid names like
+  // foo.suffix.suffix.suffix. Instead of repeating the suffix add a numeric
+  // suffix. Specifically, the clone of foo.suffix is named foo.suffix2, the
+  // clone of foo.suffix2 is named foo.suffix3 and so on.
+  const std::string dot_suffix = "." + suffix;
+  size_t index = name().rfind(dot_suffix);
+  if (index == std::string::npos) {
```
Comment (cheshire) [inline xla/hlo/ir/hlo_instruction.cc:2084]: The logic seems relatively complicated. Maybe a simple regexp match would do instead? It still seems like it should be expressible in a couple of lines, something like (in pseudocode)

```
p = r".*\.suffix(\d)$"
if (Match m = Matches(name, p)) {
  return StrCat(name.substring(m.start()), ".suffix", StrToInt(m.group(1)) + 1)
}
```

Comment (cheshire) [reply xla/hlo/ir/hlo_instruction.cc:2084]: Though seems like this code was copied, then it's fine.

**Review by cheshire — APPROVED** [review]

File: xla/hlo/ir/hlo_instruction.cc (hunk 1 above)
Comment (Tixxx) [reply xla/hlo/ir/hlo_instruction.cc:54]: i accidentally applied clang-format to the whole file. will revert this.

**Review by cheshire — APPROVED** [review]

**Pull request closed without merging**